File: yt-proxy.py
#!/usr/bin/env python3
"""Tiny YouTube search proxy — run: python3 yt-proxy.py
Serves on http://localhost:3001 with CORS enabled"""
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.request import urlopen, Request
from urllib.parse import parse_qs, urlparse, quote_plus
import json, re

class Handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET')
        self.end_headers()

    def do_GET(self):
        q = parse_qs(urlparse(self.path).query).get('q', [''])[0]
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        if not q:
            return self.wfile.write(b'[]')
        try:
            url = 'https://www.youtube.com/results?search_query=' + quote_plus(q)
            req = Request(url, headers={'User-Agent': 'Mozilla/5.0', 'Accept-Language': 'en'})
            html = urlopen(req, timeout=8).read().decode('utf-8', errors='ignore')
            m = re.search(r'var ytInitialData = (\{.*?\});', html)
            if not m:
                return self.wfile.write(b'[]')
            data = json.loads(m.group(1))
            sections = data['contents']['twoColumnSearchResultsRenderer']['primaryContents']['sectionListRenderer']['contents']
            results = []
            for sec in sections:
                for it in sec.get('itemSectionRenderer', {}).get('contents', []):
                    vr = it.get('videoRenderer')
                    if not vr or not vr.get('videoId'):
                        continue
                    results.append({
                        'id': vr['videoId'],
                        'title': (vr.get('title', {}).get('runs') or [{}])[0].get('text', ''),
                        'channel': (vr.get('ownerText', {}).get('runs') or [{}])[0].get('text', '')
                    })
            self.wfile.write(json.dumps(results[:20]).encode())
        except Exception:
            self.wfile.write(b'[]')

    def log_message(self, fmt, *args):
        print(f'[YT Proxy] {args[0]}')

print('[YT Proxy] http://localhost:3001 — ready')
HTTPServer(('', 3001), Handler).serve_forever()

File: yt-proxy.js
// Tiny YouTube search proxy — run: node yt-proxy.js
// Serves on http://localhost:3001 with CORS enabled
const http = require('http');
const https = require('https');
const PORT = 3001;

http.createServer((req, res) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET');
  if (req.method === 'OPTIONS') { res.writeHead(204); return res.end(); }

  const u = new URL(req.url, 'http://localhost');
  const q = u.searchParams.get('q');
  if (!q) { res.writeHead(400); return res.end('missing ?q='); }

  const ytUrl = 'https://www.youtube.com/results?search_query=' + encodeURIComponent(q);
  https.get(ytUrl, { headers: { 'User-Agent': 'Mozilla/5.0', 'Accept-Language': 'en' } }, ytRes => {
    let body = '';
    ytRes.on('data', c => body += c);
    ytRes.on('end', () => {
      try {
        const m = body.match(/var ytInitialData = (\{.*?\});/);
        if (!m) { res.writeHead(200, { 'Content-Type': 'application/json' }); return res.end('[]'); }
        const data = JSON.parse(m[1]);
        const sections = data.contents.twoColumnSearchResultsRenderer
          .primaryContents.sectionListRenderer.contents;
        const results = [];
        for (const sec of sections) {
          for (const it of (sec.itemSectionRenderer?.contents || [])) {
            const vr = it.videoRenderer;
            if (!vr?.videoId) continue;
            results.push({
              id: vr.videoId,
              title: vr.title?.runs?.[0]?.text || '',
              channel: vr.ownerText?.runs?.[0]?.text || ''
            });
          }
        }
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify(results.slice(0, 20)));
      } catch (e) {
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end('[]');
      }
    });
  }).on('error', () => { res.writeHead(200, { 'Content-Type': 'application/json' }); res.end('[]'); });
}).listen(PORT, () => console.log('[YT Proxy] http://localhost:' + PORT + ' — ready'));
